### PR TITLE
FeatureExtractor.predict() can now return labels in addition to features

### PR DESCRIPTION
**Pull request by cofri**
A new bool argument `return_labels` is introduced in `FeatureExtractor.predict()`. If True, the function returns a tuple `(features, labels)`. If False, the function behaves as previously: only the features are returned as a tensor.

Special cases:
- if no label is given in the dataset, `None` is returned
- if labels are one-hot encoded, they are reduced to single values.

### CODE REVIEW

File: oodeel/extractor/keras_feature_extractor.py [hunk 1]
```diff
@@ -164,22 +164,33 @@ def predict_tensor(self, tensor: TensorType) -> Union[tf.Tensor, List[tf.Tensor]
...
 
         Returns:
             List[tf.Tensor]: features
         """
         if isinstance(dataset, get_args(ItemType)):
             tensor = TFDataHandler.get_input_from_dataset_item(dataset)
-            return self.predict_tensor(tensor)
+            features = self.predict_tensor(tensor)
+            if return_labels:
+                return features, None
```
Comment (paulnovello) [inline oodeel/extractor/keras_feature_extractor.py:188]: Should be features, None if `isinstance(dataset, tf.Tensor)` and features, label otherwise. The problem is: how to identify label in dataset item? dataset[-1]? dataset['label']? (Maybe [1] to be consistent with the rest of your code)

File: oodeel/extractor/keras_feature_extractor.py [hunk 2]
```diff
@@ -190,9 +201,28 @@ def predict(
...
+            # Get labels
+            if return_labels and contains_labels:
+                lbl_batch = elem[1]  # labels must be at index 1 in the batch tuple
+                # If labels are one-hot encoded, take the argmax
+                if tf.rank(lbl_batch) > 1 and lbl_batch.shape[1] > 1:
+                    lbl_batch = tf.reshape(lbl_batch, shape=[lbl_batch.shape[0], -1])
+                    lbl_batch = tf.argmax(lbl_batch, axis=1)
+                # If labels are in two dimensions, squeeze them
+                if len(lbl_batch.shape) > 1:
+                    lbl_batch = tf.reshape(lbl_batch, [lbl_batch.shape[0]])
```
Comment (paulnovello) [inline oodeel/extractor/keras_feature_extractor.py:213]: Maybe make a function out of it and use it twice, here and in the case of  `isinstance(dataset, get_args(ItemType)) == True`

**Review by y-prudent — CHANGES_REQUESTED** [review]
Great work!
Mahalanobis code could be updated with this feature

File: oodeel/extractor/torch_feature_extractor.py [hunk 3]
```diff
@@ -224,9 +231,15 @@ def predict(
 
         if isinstance(dataset, get_args(ItemType)):
             tensor = TorchDataHandler.get_input_from_dataset_item(dataset)
-            return self.predict_tensor(tensor, detach=detach)
+            features = self.predict_tensor(tensor, detach=detach)
+            if return_labels:
```
Comment (paulnovello) [inline oodeel/extractor/torch_feature_extractor.py:235]: In that case, we should be able to also return the label found in the given item, if there is any. Should be features, None if `isinstance(dataset, tf.Tensor)` and features, label otherwise. The problem is: how to identify label in dataset item? dataset[-1]? dataset['label']? (Maybe [1] to be consistent with the rest of your code)

File: oodeel/extractor/torch_feature_extractor.py [hunk 4]
```diff
@@ -237,9 +250,28 @@ def predict(
...
+            # Get labels
+            if return_labels and contains_labels:
+                lbl_batch = elem[1]  # labels must be at index 1 in the batch tuple
+                # If labels are one-hot encoded, take the argmax
+                if len(lbl_batch.shape) > 1 and lbl_batch.shape[1] > 1:
+                    lbl_batch = lbl_batch.view(lbl_batch.size(0), -1)
+                    lbl_batch = torch.argmax(lbl_batch, dim=1)
+                # If labels are in two dimensions, squeeze them
+                if len(lbl_batch.shape) > 1:
+                    lbl_batch = lbl_batch.view([lbl_batch.shape[0]])
```
Comment (paulnovello) [inline oodeel/extractor/torch_feature_extractor.py:262]: Maybe make a function out of it and use it twice, here and in the case of  `isinstance(dataset, get_args(ItemType)) == True`

**Review by paulnovello — CHANGES_REQUESTED** [review]
See comments therein. GG for the tests!!

File: oodeel/methods/mahalanobis.py [hunk 5]
```diff
@@ -58,23 +58,7 @@ def _fit_to_dataset(self, fit_dataset: DatasetType) -> None:
...
-                _labels = self.op.reshape(_labels, [_labels.shape[0]])
-            # get features
-            _features = self.feature_extractor.predict(_images)
-            _features = self.op.flatten(_features)
-
-            labels.append(_labels)
-            features.append(_features)
-        labels = self.op.cat(labels)
-        features = self.op.cat(features)
+        features, labels = self.feature_extractor.predict(fit_dataset)
```
Comment (y-prudent) [inline oodeel/methods/mahalanobis.py:61]: You forgot the argument `return_labels=True`

**Review by y-prudent — CHANGES_REQUESTED** [review]

**Review by y-prudent — APPROVED** [review]

**Comment by cofri** [comment]
I've changed `.predict()` of the feature extractors to handle a list/tuple of tensors [data, labels]. Also added a unit test to assert this feature.
Note that the object `list` is now an `ItemType`. This is required because a PyTorch batch `next(iter(dataloader))` returns a list of `[data, labels]`, compared to TensorFlow which returns a tuple `(data, labels)`.

**Review by paulnovello — APPROVED** [review]
LGTM
